### PR TITLE
#167733643 Email verification functionality

### PR DESCRIPTION
**Pull request by daylay92**
#### What does this PR do?
Add email verification functionality to the App, ensuring a user gets a verification email after a successful registration.

#### Description of Task to be completed?
- Install SendGrid, a third-party package for sending in-app emails.
- Implement methods for signing and verifying authentication token in `helpers.js` (a utility class file).
- Implement a method for generating verification links.
- Implement a method for sending verification emails in the `mailer.js` file.
- Add the email verification method to the signup implementation.
- Refactor the routing setup of the current codebase
- Create a route and an endpoint `api/auth/verify` for validating an email verification link.
- Implement a method in the `UserService class` for updating the account status of a user as verified, in the event of a successful verification through the aforementioned route.
- Write unit tests for the functionalities mentioned above.

#### How should this be manually tested?
- Clone repo and change into the working directory
- Switch to the ft-email-verification-167733643 branch and run `npm install`
- Create a .env file and specify the variables described in the .env.sample inside this file.
- Run `npm run prestart` to create tables in your database.
- Once the database tables are created, run `npm run start:dev` to start the development server.
- Launch Postman and make a post required to the signup route `http://localhost:3000/api/auth/signup` using the field described in the swagger document of the project.
- Upon successful signup, visit your registration mail and check for an email from the app, following the instruction described in the mail.
- Upon successful verification, a JSON response should be received with updated user parameters.
- To run the unit tests, run `npm test`.
- Document feature using swagger 

#### What are the relevant pivotal tracker stories?
#167733643

#### Screenshot

test-passing
email

### CODE REVIEW

File: src/database/init.js [hunk 1]
```diff
@@ -0,0 +1,9 @@
+import db from '../models';
+import env from '../config/env-config';
+
+if (env.NODE_ENV !== 'production') {
+  const syncOptions = env.NODE_ENV === 'test' ? { force: true } : {};
+  db.sequelize.sync(syncOptions).then(() => {
+    db.sequelize.close();
+  });
+}
```
Comment (King-Success) [inline src/database/init.js]: What does this file do?

File: src/index.js [hunk 2]
```diff
@@ -1,25 +1,80 @@
...
-app.use(routes);
-app.get('/', (req, res) =>
-  res.status(200).send({ message: 'welcome to BN: jubilee-team' })
-);
-app.all('*', (req, res) => res.send({ message: 'route not found' }));
-
-app.listen(port, () => {
-  console.info(`Listening on port ${port}`);
+
+app.use(bodyParser.urlencoded({ extended: false }));
```
Comment (King-Success) [inline src/index.js]: Why need to install new packages to handle what express can handle out of the box?

File: src/migrations/20190822142128-create-user.js [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    createdAt: {
+      allowNull: false,
+      type: Sequelize.DATE
+    },
+    updatedAt: {
+      allowNull: false,
+      type: Sequelize.DATE
+    }
+  }),
+  down: (queryInterface, Sequelize) => queryInterface.dropTable('Users')
```
Comment (King-Success) [inline src/migrations/20190822142128-create-user.js]: thought we already have a user migration file?

**Review by King-Success — APPROVED** [review]

File: src/config/passport.js [hunk 4]
```diff
@@ -1,28 +1,30 @@
-const passport = require("passport");
-const LocalStrategy = require("passport-local").Strategy;
-const mongoose = require("mongoose");
-const User = mongoose.model("User");
 import dotenv from 'dotenv';
+
+const passport = require('passport');
+const LocalStrategy = require('passport-local').Strategy;
+const mongoose = require('mongoose');
```
Comment (emmanuelnwankwo) [inline src/config/passport.js:5]: I thought we have decided to remove redundant code like mongoose?

**Review by Segun-Ogundipe — APPROVED** [review]
Did a great job

**Review by dchima — APPROVED** [review]
Nice work Ayodele!

File: src/config/index.js [hunk 5]
```diff
@@ -1,7 +1,8 @@
 import dotenv from 'dotenv';
```
Comment (emmanuelnwankwo) [inline src/config/index.js:1]: Add this instead 
```suggestion
import env from './env-config';
```

**Review by emmanuelnwankwo — CHANGES_REQUESTED** [review]
Well done @daylay92. I left some comments, also check the ones @King-Success pointed out

File: src/config/passport.js (hunk 4 above)
Comment (daylay92) [reply src/config/passport.js:5]: Yes. @chis0m is working with the passport.js file to create the social provider authentication is I left the boiler plate for him

Comment (emmanuelnwankwo) [reply src/config/passport.js:5]: Okay

**Review by emmanuelnwankwo — APPROVED** [review]

File: package.json [hunk 6]
```diff
@@ -4,10 +4,10 @@
...
   "main": "index.js",
   "scripts": {
-    "test": "npm run initdb:test && cross-env NODE_ENV=test  nyc --reporter=text --reporter=html mocha --timeout 10000 -c --require @babel/register --exit",
+    "test": "npm run migrate:reset && npm run migrate && cross-env NODE_ENV=test nyc --reporter=text --reporter=html mocha --timeout 10000 -c --require @babel/register --exit",
     "coveralls": "nyc --reporter=lcov --reporter=text-lcov npm test",
-    "migrate:reset": "babel-node ./node_modules/.bin/sequelize db:migrate:undo:all",
-    "migrate": "babel-node ./node_modules/.bin/sequelize db:migrate",
+    "migrate:reset": "cross-env NODE_ENV=test sequelize db:migrate:undo:all",
+    "migrate": "cross-env NODE_ENV=test sequelize db:migrate",
     "start:dev": "npm run initdb && nodemon --exec babel-node ./src/index.js",
```
Comment (Baystef) [inline package.json]: `initdb` script doesn't exist anymore, kindly update

**Review by Baystef — APPROVED** [review]
Great job

**Review by Baystef — APPROVED** [review]

**Review by chis0m — APPROVED** [review]

File: src/index.js (hunk 2 above)
Comment (daylay92) [reply src/index.js]: It's been taken out on your request

File: src/database/init.js (hunk 1 above)
Comment (daylay92) [reply src/database/init.js]: It's been removed. It was formerly used to force sync the test DB upon migrations, but as discussed I have implemented a better solution for that and it's been removed.

**Pull request merged**